src/Symfony/Component/Mime/MimeTypes.php

```diff
@@ -0,0 +1,3130 @@
+ * previously registered ones.
+ *
+ * Re-registering a built-in guesser also allows you to configure it:
+ *
+ *     $guesser = new MimeTypes();
+ *     $guesser->registerGuesser(new FileinfoMimeTypeGuesser('/path/to/magic/file'));
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class MimeTypes implements MimeTypesInterface
+{
+    private $guessers = [];
+    private static $default;
+
+    public function __construct(array $map = [])
+    {
+        foreach ($map as $mimeType => $extensions) {
+            self::$extensions[$mimeType] = $extensions;
+        }
+        $this->registerGuesser(new FileBinaryMimeTypeGuesser());
+        $this->registerGuesser(new FileinfoMimeTypeGuesser());
+    }
+
+    public static function setDefault(self $default)
+    {
+        self::$default = $default;
+    }
+
+    public static function getDefault(): self
+    {
+        return self::$default ?? self::$default = new self();
+    }
+
+    /**
+     * Registers a MIME type guesser.
+     *
+     * The last registered guesser has precedence over the other ones.
+     */
+    public function registerGuesser(MimeTypeGuesserInterface $guesser)
+    {
+        array_unshift($this->guessers, $guesser);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getExtensions(string $mimeType): array
+    {
+        return self::$map[$mimeType] ?? [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMimeTypes(string $ext): array
+    {
+        return self::$reverseMap[$ext] ?? [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isGuesserSupported(): bool
+    {
+        foreach ($this->guessers as $guesser) {
+            if ($guesser->isGuesserSupported()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * The file is passed to each registered MIME type guesser in reverse order
+     * of their registration (last registered is queried first). Once a guesser
+     * returns a value that is not null, this method terminates and returns the
+     * value.
+     */
+    public function guessMimeType(string $path): ?string
+    {
+        foreach ($this->guessers as $guesser) {
+            if (!$guesser->isGuesserSupported()) {
+                continue;
+            }
+
+            if (null !== $mimeType = $guesser->guessMimeType($path)) {
+                return $mimeType;
+            }
+        }
+
+        if (!$this->isGuesserSupported()) {
+            throw new \LogicException('Unable to guess the MIME type as no guessers are available (have you enable the php_fileinfo extension?).');
+        }
+
+        return null;
+    }
+
+    /**
+     * A map of MIME types and their default extensions.
+     *
+     * Updated from upstream on 2019-01-16
+     *
+     * @see Resources/bin/update_mime_types.php
+     */
+    private static $map = [
+        'application/acrobat' => ['pdf'],
```

Comment:
Does this kind of list scale? Even currently browser has trouble opening this file in github. It's almost 2000 entries now. Implementation of updating it could also be simpler if it was in external file and just var_export used. This point was raised by somebody in Slack originally.

Reply:
Scale in what sense? Displaying it on Github is not a goal. PHP manages it fairly well and is even able to optimize it.

Reply:
Goal should be scale it for both PHP and human reviewing the core functionality of MimeTypes.php. Currently this list is mostly just noise for human since it's something generated. Thanks to Opcache it should be fairly good optimized with external file as well, shouldn't it?